src/modelfree/transparent.py

```diff
@@ -0,0 +1,188 @@
+        action = self.deterministic_action if deterministic else self.action
+        outputs = [action, self.value_flat, self.state_out, self.neglogp, self.ff_out]
+        feed_dict = self._make_feed_dict(obs, state, mask)
+        a, v, s, neglogp, ff = self.sess.run(outputs, feed_dict)
+        state = []
+        for x in s:
+            state.append(x.c)
+            state.append(x.h)
+        state = np.array(state)
+        state = np.transpose(state, (1, 0, 2))
+
+        # right now 'hid' only supports hidden state of policy
+        # which is the last of the four state vectors
+        transparent_objs = (obs, np.concatenate(ff['policy']),
+                            np.concatenate(ff['value']), state[:, -1, :])
+        transparency_dict = self.get_transparency_dict(transparent_objs)
+        return a, v, state, neglogp, transparency_dict
+
+
+class TransparentMlpPolicyValue(TransparentPolicy, MlpPolicyValue):
+    """gym_compete MlpPolicyValue which is also transparent."""
+    def __init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch, transparent_params,
+                 hiddens=None, scope="input", reuse=False, normalize=False):
+        MlpPolicyValue.__init__(self, sess, ob_space, ac_space, n_env, n_steps, n_batch,
+                                hiddens=hiddens, scope=scope, reuse=reuse, normalize=normalize)
+        TransparentPolicy.__init__(self, transparent_params)
+
+    def get_obs_sizes(self):
+        return self.ob_space.shape[0], sum(self.hiddens), sum(self.hiddens), None
+
+    def step(self, obs, state=None, mask=None, deterministic=False):
+        action = self.deterministic_action if deterministic else self.action
+        outputs = [action, self.value_flat, self.neglogp, self.ff_out]
+        a, v, neglogp, ff = self.sess.run(outputs, {self.obs_ph: obs})
+
+        transparent_objs = (obs, np.concatenate(ff['policy']), np.concatenate(ff['value']), None)
+        transparency_dict = self.get_transparency_dict(transparent_objs)
+        return a, v, self.initial_state, neglogp, transparency_dict
+
+
+class TransparentCurryVecEnv(CurryVecEnv):
+    """CurryVecEnv that provides transparency data about its policy in its infos dicts."""
+    def __init__(self, venv, policy, agent_idx=0):
+        super().__init__(venv, policy, agent_idx)
+        self.underlying_policy = policy.policy
+        if not isinstance(self.underlying_policy, TransparentPolicy):
+            raise TypeError("Error: policy must be transparent")
+        self._action = None
+
+        obs_aug_amount = self.underlying_policy.get_obs_aug_amount()
+        if obs_aug_amount > 0:
+            obs_aug_space = gym.spaces.Box(-np.inf, np.inf, (obs_aug_amount,))
+            self.observation_space = _tuple_space_augment(self.observation_space, agent_idx,
+                                                          augment_space=obs_aug_space)
+
+    def step_async(self, actions):
```

Comment:
Why are we evaluating the policy in `step_wait` now rather than `step_async`? I don't see why it should be different to `CurryVecEnv`. Maybe there's an argument for doing it in `step_wait` so `step_async` won't block, but then we should update `CurryVecEnv` too.

Reply:
We need to do it in `step_wait` and `reset` because we may want to augment the observation of the non-curried agent. Thus, we have to get a fresh `self._data` before we return observations

Reply:
^ That's no longer true, since we're scrapping the whole observation augmentation thing now. I'll put it back in step_async